nn/conformer.py

```diff
@@ -0,0 +1,205 @@
+"""
+Conformer code.
+Ref: https://arxiv.org/abs/2005.08100
+"""
+
+from typing import Tuple, List, Union
+
+from . import Module, ModuleList, LayerRef, Linear, dropout, layer_norm, batch_norm, Conv, swish, glu, split_dims, \
+  merge_dims, pool
+
+
+class _PositionwiseFeedForward(Module):
+  """
+  Conformer position-wise feedforward neural network layer
+      FF -> Activation -> Dropout -> FF
+  """
+
+  def __init__(self, d_model: int, d_ff: int, dropout: float, activation, l2: float = 0.0):
+    """
+    :param d_model:
+    :param d_ff:
+    :param dropout:
+    :param activation:
+    :param l2:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+    self.activation = activation
+
+    self.linear1 = Linear(n_out=d_ff, l2=l2)
+    self.linear2 = Linear(n_out=d_model, l2=l2)
+
+  def forward(self, inp: LayerRef) -> LayerRef:
+    return self.linear2(dropout(self.activation(self.linear1(inp)), dropout=self.dropout))
+
+
+class _ConformerConvBlock(Module):
+  """
+  Conformer convolution block
+      FF -> GLU -> depthwise conv -> BN -> Swish -> FF
+  """
+
+  def __init__(self, d_model: int, kernel_size: Tuple[int], l2: float = 0.0):
+    """
+    :param d_model:
+    :param kernel_size:
+    :param l2:
+    """
+    super().__init__()
+
+    self.positionwise_conv1 = Linear(n_out=d_model * 2, l2=l2)
+    self.depthwise_conv = Conv(n_out=d_model, filter_size=kernel_size, groups=d_model, l2=l2, padding='same')
+    self.positionwise_conv2 = Linear(n_out=d_model, l2=l2)
+
+  def forward(self, inp: LayerRef) -> LayerRef:
+    x_conv1 = self.positionwise_conv1(inp)
+    x_act = glu(x_conv1)
+    x_depthwise_conv = self.depthwise_conv(x_act)
+    x_bn = batch_norm(x_depthwise_conv)
```

Comment:
I just realized that `batch_norm` was buggy. It has variables. So you should use the module to make the variable naming consistent.